### PR TITLE
Use GetVersion() API instead of ver command

### LINKED ISSUE #55083: Should use GetVersion vs Shelling out to ver
https://github.com/kubernetes/kubernetes/blob/d837a6a2eabee7cc3ed2dce1d031a1a491d8faec/pkg/kubelet/winstats/perfcounter_nodestats.go#L54

Using GetVersion() will likely be more reliable/faster/better than shelling out to ver... if you need more data than GetVersion you might also follow the pattern from Moby https://github.com/moby/moby/blob/54251b53d7081f1471c0a2cb23fd4e198e71bd14/pkg/parsers/kernel/kernel_windows.go

@kubernetes/sig-windows

### PR DESCRIPTION
**Pull request by feiskyer**
**What this PR does / why we need it**:

Should use GetVersion vs Shelling out to ver.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #55083

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by feiskyer** [comment]
Note that with this the kernel verson is in different format as before, e.g. on my local windows 10 laptop

- `ver` command would output ` 10.0.16299.19`
- while GetVersion would got `16299.15.amd64fre.rs3_release.170928-1534`

/cc taylorb-microsoft is this what we expected?

**Comment by taylorb-microsoft** [comment]
@feiskyer - thanks for coding this up so quickly!

 - For windows versions the kernel matches usermode so there is only one version.
 - Explanation of fields from ver:
   - The first two fields (10.0) specific the OS Major version, which is unlikely to change but you can get it from GetVersionEX (https://msdn.microsoft.com/en-us/library/windows/desktop/ms724833(v=vs.85).aspx)
   - The third field (16299) is the buildNumber - that number is important as it increments with every full build.
    - The fourth field (19) is the UBR number, which is really what is required from the registry.  The UBR number specifies what updates have been applied, each new Windows update increments this number.
 - Expiation of fields from BuildLabEx
   - First field is the (16299) is the buildNumber (you can also query the CurrentBuildNumber key vs parsing)
    - Second field is the UBR number ( you can also query the UBR key vs parsing)
    - Third field is the architecture
    - Forth field is the build Branch ( you can also query the BuildBranch key vs parsing)
    - Fifth field is the time stamp (not overly helpful unless your part of the Windows team)

**Comment by taylorb-microsoft** [comment]
To your question on why ver is returning a different URB than what is in the BuildLabEx registry key I'm not sure - can you see what its in that key vs the UBR key?  The ver code queries the UBR key for that information.

Specifically what they are doing is:
1)  Call GetVersion() (https://msdn.microsoft.com/en-us/library/windows/desktop/ms724439(v=vs.85).aspx)
2)  Call an local GetVersionRevision() API - which uses the UBR registry key
3)  Format a string:
`           "%d.%d.%05d.%d", 
           (ver & 0xFF), 
           ((ver >> 8)   & 0xFF),
           ((ver >> 16) & 0xFFFF),
           revision)
`

**Comment by feiskyer** [comment]
@taylorb-microsoft Addressed comments. PTAL

**Comment by JiangtianLi** [comment]
@feiskyer @taylorb-microsoft  Thanks for fixing this. I have a similar fix in #55496 but will revert to use yours. One question, for Linux, 
```
OS-IMAGE                              KERNEL-VERSION
Debian GNU/Linux 8 (jessie)           4.11.0-1013-azure
```
Should the osImageVersion be something more verbal such as "Microsoft Windows 10 Enterprise" or "Microsoft Windows Server Datacenter", e.g. HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProductName? Just my 2 cents.

**Comment by feiskyer** [comment]
@JiangtianLi Reasonable, updated with the value from `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\ProductName`. PTAL

**Comment by JiangtianLi** [comment]
@feiskyer LGTM. Thanks for addressing it.

**Comment by feiskyer** [comment]
Kindly ping @tallclair @yujuhong

**Comment by feiskyer** [comment]
ping  @tallclair @dchen1107 Could you take a look at this PR?

File: pkg/kubelet/winstats/version.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		0,
+		windows.KEY_READ,
+		&h); err != nil {
+		return 0, err
+	}
+	defer windows.RegCloseKey(h)
+
+	var buf [1 << 10]uint16
+	var typ uint32
+	n := uint32(len(buf) * 2) // api expects array of bytes, not uint16
```
Comment (tallclair) [inline pkg/kubelet/winstats/version.go]: nit: prefer
```
n := uint32(len(buf) * unsafe.Sizeof(uint16))
```

File: pkg/kubelet/winstats/version.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	if err := windows.RegOpenKeyEx(windows.HKEY_LOCAL_MACHINE,
+		windows.StringToUTF16Ptr(`SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\`),
+		0,
+		windows.KEY_READ,
+		&h); err != nil {
+		return 0, err
+	}
+	defer windows.RegCloseKey(h)
+
+	var buf [1 << 10]uint16
```
Comment (tallclair) [inline pkg/kubelet/winstats/version.go]: nit: does buf need to be that large? What are typical values of this key?

File: pkg/kubelet/winstats/version.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return "", err
+	}
+
+	revision, err := getVersionRevision()
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%d.%d.%05d.%d\n", (ver & 0xFF), ((ver >> 8) & 0xFF), ((ver >> 16) & 0xFFFF), revision), nil
```
Comment (tallclair) [inline pkg/kubelet/winstats/version.go]: these versions look backwards to me? might be more readable if you break it out like so:
```go
var (
  major = ...
  minor = ...
  etc
)
```

File: pkg/kubelet/winstats/version.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// getOSImageVersion gets the osImage name and version.
+func getOSImageVersion() (string, error) {
```
Comment (tallclair) [inline pkg/kubelet/winstats/version.go]: extract the common code to a helper: `func getCurrentVersionVal(key string) (string, error)`

File: pkg/kubelet/winstats/version.go (hunk 2 above)
Comment (feiskyer) [reply pkg/kubelet/winstats/version.go]: It's the value of windows registy, may includes integers, uint arrays or string.

File: pkg/kubelet/winstats/version.go (hunk 3 above)
Comment (feiskyer) [reply pkg/kubelet/winstats/version.go]: ack, good catch

File: pkg/kubelet/winstats/version.go (hunk 4 above)
Comment (feiskyer) [reply pkg/kubelet/winstats/version.go]: ack

**Comment by feiskyer** [comment]
@tallclair Rebased and addressed comments. PTAL

File: pkg/kubelet/winstats/version.go (hunk 2 above)
Comment (tallclair) [reply pkg/kubelet/winstats/version.go]: Generally, yes registry values could be any of those. But in this specific case, you must have some expectations for what the 2 fields you're querying can hold? E.g. if it's a version string it's probably not more than, what, 32 characters?  Allocating 2KB everytime this is called seems like kind of a lot.

File: pkg/kubelet/winstats/version.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	revision, err := getVersionRevision()
+	if err != nil {
+		return "", err
+	}
+
+	major := ver & 0xFF
+	minor := (ver >> 8) & 0xFF
+	build := (ver >> 16) & 0xFFFF
```
Comment (tallclair) [inline pkg/kubelet/winstats/version.go]: So if I'm reading this correctly, the value `0xABCDEF` would be parsed as:
```
major = EF
minor = CD
build = AB
```
Is that correct? Or should it be the other way around?

Comment (feiskyer) [reply pkg/kubelet/winstats/version.go]: @tallclair right. A tipical example is 0x23F00206, represents 6.2.9200.

File: pkg/kubelet/winstats/version.go (hunk 2 above)
Comment (feiskyer) [reply pkg/kubelet/winstats/version.go]: Good catch. I see all other places using this API are allocating such size, but it is not required such big in this package.

**Comment by feiskyer** [comment]
@tallclair Addressed comments. PTAL

**Comment by feiskyer** [comment]
@tallclair Could you also help to approve the milestone?

**Pull request merged**